utcp.c

```diff
@@ -587,58 +732,88 @@ static void ack(struct utcp_connection *c, bool sendatleastone) {
     pkt->hdr.ctl = ACK;
     pkt->hdr.aux = 0;
 
+    int err = 0;
     do {
         uint32_t seglen = left > c->utcp->mtu ? c->utcp->mtu : left;
         uint32_t bufpos = seqdiff(c->snd.nxt, c->snd.una);
         pkt->hdr.seq = c->snd.nxt;
 
         left -= seglen;
 
+        // adjust packet data length for the segment length
         // when FIN is not ack'ed yet len must be at least 1
+        size_t datalen = seglen;
         if(seglen && fin_wanted(c, c->snd.nxt + seglen)) {
-            seglen--;
+            datalen--;
             pkt->hdr.ctl |= FIN;
         }
 
-        buffer_copy(&c->sndbuf, pkt->data, bufpos, seglen);
+        buffer_copy(&c->sndbuf, pkt->data, bufpos, datalen);
 
-        size_t pktlen = sizeof pkt->hdr + seglen;
+        size_t pktlen = sizeof pkt->hdr + datalen;
         print_packet(c->utcp, "send", pkt, pktlen);
-        if(c->utcp->send(c->utcp, pkt, pktlen) != pktlen) {
-            debug("Error: utcp ack() failed to send pkt %u with ack %u", pkt->hdr.seq, pkt->hdr.ack);
-            // break loop and hope to recover so time later
-            break;
+        ssize_t sent = c->utcp->send(c->utcp, pkt, pktlen);
+        if(sent != pktlen) {
+            if(sent > pktlen) {
+                utcp_log_send_error(pkt, pktlen, sent, false);
+            }
+            else if(sent >= 0 || sent == UTCP_WOULDBLOCK) {
+                // when no data could be sent with possibly the header broken
+                // or when the socket would block, don't advance but retry later
+                utcp_log_send_error(pkt, pktlen, sent, false);
+                err = UTCP_WOULDBLOCK;
+                break;
+            }
+            else {
+                // the pkt receiver might have gone offline causing the routing to fail
+                // break loop and hope to recover some time later
+                // it would only cause a retransmit when skipped
+                utcp_log_send_error(pkt, pktlen, sent, false);
+                err = UTCP_ERROR;
+                break;
+            }
         }
 
+        // if anything sent, andvance
+        c->snd.nxt += seglen;
```

Comment:
the snd.nxt sequence is incremented by the seglen, that is correct since the FIN flags takes a virtual size of 1 for the transmit, the actual packet datalen however doesn't